Fix optional headers

gfontenot opened the pull request:
- Improve header parsing for fewer false positives
- Use `option` properly (with `try`) so that the default is returned if
  parsing fails.

gfontenot committed:
Improve the parsing of header names

Just specifying that we shouldn't see a ":" or a newline isn't actually
enough here. What we _actually_ want is all characters up to ":" without
any newlines. That will match properly, and ensure that we're parsing
our headers correctly.

gfontenot committed:
Allow header parsing to fail

We don't always have headers, and that's OK. If we don't we should just
go on with our day.

If you don't use `try` with `option`, then the parser always consumes
content, and so `option` never returns the default value. By using
`try`, we can ensure that if the parser fails it does _not_ consume
content and so the default can be returned.

pbrisbin commented on src/Reflow/Parser.hs:
```diff
@@ -31,7 +31,10 @@ header = do
 headerName :: Parser Text
-headerName = pack <$> (many1 $ noneOf ":\n")
+headerName = do
+    n <- manyTill (noneOf "\n") (lookAhead $ string ":")
```
I thought header names had to be in a limited character range. If so, this could be something like:

``` hs
fmap pack $ (<>)
    <$> many1 (oneOf '-':['0'..'Z'])
    <*> string ":"
```

Do you actually want the trailing `":"` in the header name? If not, this could shortened to:

``` hs
many1 (oneOf $ '-':['0'..'Z']) <* string ":"
```

gabebw replied:
How is this different from  `many1 $ noneOf "\n:"`?

jferris replied:
Do you need the `lookAhead` here?

gfontenot replied:
so many questions!

@pbrisbin I _do_ want the `:` here. I'm not actually caring about splitting the header into name/value pairs, I just care about recognizing them and then reconstructing the line. I do like the idea of using the specific character set though, I'll make that change.

@gabebw `noneOf "\n:"` will consume input till it sees either `:` _or_ `\n`. This isn't actually what we want though, since we want to make sure we're parsing strings that look like `"foo: bar\n"` and _not_ `"f\noo: bar"`.

@jferris My understanding is that `lookAhead` is needed because without it, we'll consume and then throw away the trailing ":". I suppose I could re-add it manually though?

jferris commented on src/Reflow/Parser.hs:
```diff
@@ -31,7 +31,10 @@ header = do
+headerName = do
+    n <- manyTill (noneOf "\n") (lookAhead $ string ":")
+    s <- string ":"
+    return $ pack $ n <> s
```
You can use applicative if you'd like:

``` haskell
pack . mappend <$> manyTill (noneOf "\n") (lookAhead $ string ":") <*> string ":"
```

gfontenot replied:
Ah, yes, _that's_ what this should look like. I could _not_ figure out how this was supposed to look with applicative style. If I consume the `:`, this could be even nicer:

``` haskell
pack . mappend <$> manyTill (noneOf "\n") (string ":") <*> pure ":"
```

gfontenot committed:
Improve header parsing _even more_

Headers actually have a strict set of allowed characters. This means we
can be even more explicit with our header name parsing, which means our
header parsing gets stricter overall. We can also use applicative style
here to remove some of the duplication and reconstruction.
